### PR TITLE
Don't set shutdownDoneCh to nil

### PR DESCRIPTION
**Pull request by averche**
I had a chat with @mpalmi about #18358. I think his solution should be perfectly safe and should work. 

This is an optional improvement that just makes the code a bit more clear. Instead of setting the channel to nil, I added a `bool` to check whether it's been closed. Reading from a closed channel in `ShutdownWait()` will return immediately.

I don't really know how to test this and I'm not familiar with the code-base. If this change doesn't make sense or is too high risk, I'm totally fine withdrawing the PR :)

### CODE REVIEW

File: vault/core.go [hunk 1]
```diff
@@ -1339,10 +1336,9 @@ func (c *Core) Shutdown() error {
 	c.stateLock.Lock()
 	defer c.stateLock.Unlock()
 
-	doneCh := c.shutdownDoneCh.Load().(chan struct{})
-	if doneCh != nil {
-		close(doneCh)
-		c.shutdownDoneCh.Store((chan struct{})(nil))
+	if !c.shutdownDoneChClosed {
```
Comment (ncabatoff) [inline vault/core.go:1339]: Isn't this racy? What if two calls to Shutdown are made at the same time?  Then we could close twice and panic.

**Comment by ncabatoff** [comment]
Can you expand on how this makes things clearer?  I think the "set a closed channel to nil" pattern is pretty common in our codebase (and in Go in general?), so I'm not sure this is an improvement, consistency-wise.

File: vault/core.go (hunk 1 above)
Comment (averche) [reply vault/core.go:1339]: This is protected by the `stateLock` mutex above, so the second call to shutdown would always find `shutdownDoneChClosed == true`

**Comment by averche** [comment]
> Can you expand on how this makes things clearer? I think the "set a closed channel to nil" pattern is pretty common in our codebase (and in Go in general?), so I'm not sure this is an improvement, consistency-wise.

To me, setting a closed channel to `nil` seems like an anti-pattern, since the channel receiver(s) now need to:
1. check if the channel is nil before receiving from it
2. add extra synchronization to ensure that the nil check is thread-safe

In particular, in the previous implementation of the `ShutdownWait()` function, the nil check was not strictly thread-safe:

```go
func (c *Core) ShutdownWait() error {
	donech := c.ShutdownDone()
	err := c.Shutdown()
	if donech != nil {
		<-donech            // can theoretically end receive from a nil channel
	}
	return err
}
```

With the new implementation, reading from a closed channel is always safe and will simply return immediately. If the channel is not-yet-closed, reading `<-donech` will block until it is closed.

File: vault/core.go (hunk 1 above)
Comment (ncabatoff) [reply vault/core.go:1339]: True!

**Comment by ncabatoff** [comment]
> To me, setting a closed channel to nil seems like an anti-pattern, since the channel receiver(s) now need to:

No, the channel receiver shouldn't be the one to look to see if it's nil.  For one thing, channels are pointers, and generally speaking the reader and writer aren't going to be looking at the same variable.  But the purpose of setting a channel nil is to prevent double-closing it and causing a panic, and only senders should close channels.  A reader shouldn't care if a channel is nil, since all they can do with a channel is read from it, and a nil channel can be read from, it just blocks forever.  But since we closed the channel before setting it to nil, the reader will have got an empty read from the closed channel, and it should have its own copy of the channel pointer anyway.

I've been speaking so far about the pattern in general, but I recognize that this situation is a little different.  Here we didn't pass the channel to a reader so that it could create its own copy of the pointer, instead we're storing it in core, and the reader is calling a method that fetches the same channel/pointer as the channel "writer" (closer) uses.

I'm wondering if a channel is really the right tool for the job here.  Wouldn't a waitgroup be more suitable?  Create it and `Add(1)` to it when the core is created.  Call `Done` on it where currently we close the channel.  ShutdownWait just does `c.Shutdown(); c.shutdownWg.Wait()`.

**Comment by averche** [comment]
> For one thing, channels are pointers, and generally speaking the reader and writer aren't going to be looking at the same variable.

This is true. If the reader and writer are looking at different copies of the channel pointer then setting the writer's channel to nil should be fine. I believe this actually applies to `ShutdownWait()` since `donech := c.ShutdownDone()`  makes a local copy of the channel pointer. So I was wrong -- `ShutdownWait()` is thread safe. Still this was not easy to ascertain. 

>  A reader shouldn't care if a channel is nil, since all they can do with a channel is read from it, and a nil channel can be read from, it just blocks forever. 

I don't think that's usually the desired behavior. If the reader will sometimes block forever, this could cause an unintended deadlock.

> But since we closed the channel before setting it to nil, the reader will have got an empty read from the closed channel, and it should have its own copy of the channel pointer anyway.

Since the reader is often going to be on a different goroutine than the writer, it could start reading before or after the channel is closed/set to nil. If it does not have its own copy of the channel pointer, it could lead to bugs. 

> I'm wondering if a channel is really the right tool for the job here. Wouldn't a waitgroup be more suitable? Create it and Add(1) to it when the core is created. Call Done on it where currently we close the channel. ShutdownWait just does c.Shutdown(); c.shutdownWg.Wait().

I think a `WaitGroup` would work just as well. However, It would also need a similar synchronization mechanism because calling `wg.Done()` twice [will cause a panic](https://go.dev/play/p/Z0tdswd9zGj). I slightly prefer a channel here to keep `ShutdownDone() <-chan struct{}` backwards compatible (though it's not supposed to be used externally).

**Comment by averche** [comment]
Closing this for now since setting channel to nil after closing is idiomatic in Vault. I can reopen if this approach is preferred.

**Pull request closed without merging**